File: q76-100/q88.py
# 2020.12.30
# Problem Statement:
# https://leetcode.com/problems/merge-sorted-array/

class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        
        # nums1_ptr will always points to the position in num2 to be compared with the current to-be-inserted element in nums2
        nums1_ptr = 0
        for i in range(0, n):
            # not yet the insert position, increase nums1_ptr
            while nums2[i] > nums1[nums1_ptr] and nums1_ptr <= m+i-1:
                nums1_ptr = nums1_ptr + 1
            # do the insertion
            nums1.insert(nums1_ptr, nums2[i])
            # delete the last item
            nums1.pop()
        
        return

File: Amazon/q88.java
// 2022.06.22
// Problem Statement:
// https://leetcode.com/problems/merge-sorted-array/

// idea: first move every 'effective' elements to the right inside nums1
// then merge nums1 and nums2
class Solution {
    public void merge(int[] nums1, int m, int[] nums2, int n) {
        for (int i=m-1; i>=0; i--) {
            int temp = nums1[i];
            nums1[i] = 0;
            nums1[nums1.length-m+i] = temp;
        }
        
        // the start comparing position for nums1 and nums2
        int idx_1=nums1.length-m, idx_2 = 0;
        for (int i=0; i<nums1.length; i++) {
            if (idx_1>=nums1.length) { // nums1 is done
                nums1[i] = nums2[idx_2];
                idx_2++;
            } else if (idx_2>=nums2.length) { // nums2 is done
                nums1[i] = nums1[idx_1];
                idx_1++;
            } else if (nums1[idx_1]<=nums2[idx_2]) {
                nums1[i] = nums1[idx_1];
                idx_1++;
            } else {
                nums1[i] = nums2[idx_2];
                idx_2++;
            }
        }
        return;
    }
}